### PR TITLE
[Speculative Decoding] Fixing hidden states handling in batch expansion

### LINKED ISSUE #7505: [Bug]: Error in how HiddenStates are handled for speculative decoding
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

In draft models like Medusa, MLPSpeculator etc., when spec. decode is disabled (e.g. when the num_tokens + spec_tokens > max_len of the model) HiddenStates are not handled properly which causes an invalid shape error.

How to reproduce?

Code:
```python
from vllm import LLM, SamplingParams

llm = LLM(
    model="JackFram/llama-160m",
    speculative_model="ibm-fms/llama-160m-accelerator",
    num_speculative_tokens=3,
    use_v2_block_manager=True,
    enforce_eager=True,
)

prompt = "The president of the United States is"

output = llm.generate(prompt, SamplingParams(max_tokens=2048, ignore_eos=True))
```

Output:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-1-dfa52d56a4c5>](https://localhost:8080/#) in <cell line: 12>()
     10 prompt = "The president of the United States is"
     11 
---> 12 output = llm.generate(prompt, SamplingParams(max_tokens=2048, ignore_eos=True))

10 frames
[/usr/local/lib/python3.10/dist-packages/vllm/spec_decode/spec_decode_worker.py](https://localhost:8080/#) in _verify_tokens(self, seq_group_metadata_list, proposal_scores, proposals, max_proposal_len)
    645             # Contract hidden states based on accepted tokens
    646             hs_size = hidden_states.shape[1]
--> 647             hidden_states = hidden_states.reshape(-1, max_proposal_len + 1,
    648                                                   hs_size)
    649             accepted_index = accepted_token_ids + 1  # Convert -1 to 0

RuntimeError: shape '[-1, 4, 768]' is invalid for input of size 768
```

Code:
```python
from vllm import LLM, SamplingParams

llm = LLM(
    model="JackFram/llama-68m",
    speculative_model="abhigoyal/vllm-medusa-llama-68m-random",
    num_speculative_tokens=3,
    use_v2_block_manager=True,
    enforce_eager=True,
)

prompt = "The president of the United States is"

output = llm.generate(prompt, SamplingParams(max_tokens=2048, ignore_eos=True))
```

Output:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-1-415db326cfe4>](https://localhost:8080/#) in <cell line: 12>()
     10 prompt = "The president of the United States is"
     11 
---> 12 output = llm.generate(prompt, SamplingParams(max_tokens=2048, ignore_eos=True))

10 frames
[/usr/local/lib/python3.10/dist-packages/vllm/spec_decode/spec_decode_worker.py](https://localhost:8080/#) in _verify_tokens(self, seq_group_metadata_list, proposal_scores, proposals, max_proposal_len)
    645             # Contract hidden states based on accepted tokens
    646             hs_size = hidden_states.shape[1]
--> 647             hidden_states = hidden_states.reshape(-1, max_proposal_len + 1,
    648                                                   hs_size)
    649             accepted_index = accepted_token_ids + 1  # Convert -1 to 0

RuntimeError: shape '[-1, 4, 768]' is invalid for input of size 768
```

### PR DESCRIPTION
**Pull request by abhigoyal1997**
This PR fixes the handling of hidden_states in batch expansion.

Fix #7505

### CODE REVIEW

**Comment by cadedaniel** [comment]
* can you add a test?
* what exactly does this fix? were the hidden states applied incorrectly?

**Comment by abhigoyal1997** [comment]
> * can you add a test?
Would an E2E test be sufficient? I can add the example I mentioned in the issue where we see the error.

> * what exactly does this fix? were the hidden states applied incorrectly?
There was an error in a specific scenario. Consider sequences for which spec. decode is disabled (e.g., when num_tokens + spec_tokens > max_model_len). For such sequences, proposal length is set to 0 and they are handled separately from other sequences. Then when contracting the batch (in `BatchExpansionTop1Scorer._contract_batch`), we merge spec and non_spec sequences into shape (batch_size, spec_length + 1). While this is done correctly for other tensors, hidden_states were kept unchanged before and only reshaped later on: https://github.com/vllm-project/vllm/blob/fc93e5614374688bddc432279244ba7fbf8169c2/vllm/spec_decode/spec_decode_worker.py#L650

This reshaping works fine if all sequences have proposal length > 0, but if for any sequence spec decode is disabled, it doesn't work. This PR changes batch expansion (mainly `BatchExpansionTop1Scorer._contract_batch`) such that hidden_states are manipulated and arranged like other tensors like tokens, probs and logprobs. This ensures hidden_states are correctly updated in all cases.

**Comment by cadedaniel** [comment]
Thanks for the writeup.

> Would an E2E test be sufficient? I can add the example I mentioned in the issue where we see the error.

An E2E test is good. I think the issue here is that we should validate an accuracy level, not just that it doesn't crash. For this you can build on this PR https://github.com/vllm-project/vllm/pull/6454 which adds an assertion that draft acceptance rate is 100%.

**Comment by abhigoyal1997** [comment]
> An E2E test is good. I think the issue here is that we should validate an accuracy level, not just that it doesn't crash. For this you can build on this PR #6454 which adds an assertion that draft acceptance rate is 100%.

Makes sense, but any draft model which uses hidden states can't be identical to the target model. In that case I am not sure if we can get draft acceptance rate of 100%. What do you suggest then?

**Comment by cadedaniel** [comment]
You can run it for the test prompts and record the accuracy, then assert it doesn’t go below that fixed value plus some epsilon.

**Comment by abhigoyal1997** [comment]
@cadedaniel I've added the test along with other MLPSpeculator tests. Can you please review? Thanks!

**Review by njhill — APPROVED** [review]
Thanks @abhigoyal1997!

**Comment by njhill** [comment]
@cadedaniel you good with this being merged?

File: tests/spec_decode/e2e/test_mlp_correctness.py [hunk 1]
```diff
@@ -82,6 +82,48 @@ def test_mlp_e2e_greedy_correctness(baseline_llm_generator, test_llm_generator,
...
+    """Verify acceptance rate with different batch size and large output 
+    length."""
+    run_equality_correctness_test(baseline_llm_generator,
+                                  test_llm_generator,
+                                  batch_size,
+                                  max_output_len=output_len,
+                                  temperature=0.0,
+                                  seeded=True,
+                                  force_output_len=True,
+                                  expected_acceptance_rate=0.48)
```
Comment (cadedaniel) [inline tests/spec_decode/e2e/test_mlp_correctness.py:124]: @njhill do you know what value we should expect here?

**Review by cadedaniel — APPROVED** [review]
yep, good to merge! thanks for the great testing here

File: tests/spec_decode/e2e/test_mlp_correctness.py (hunk 1 above)
Comment (cadedaniel) [reply tests/spec_decode/e2e/test_mlp_correctness.py:124]: (like, is this ballpark correct)

Comment (njhill) [reply tests/spec_decode/e2e/test_mlp_correctness.py:124]: @cadedaniel the value looks reasonable but I’m actually not sure what’s expected, will try to find out when I get a chance.

**Pull request merged**